### ISSUE TITLE
[Qt] Multiplayer Window hangs on Linux

### ISSUE THREAD

**Issue by shinyquagsire23** [mgba-emu/mgba#144]
On both the latest build and 0.3.0 I found that the Multiplayer Window doesn't work for the most part, unfortunately. I tried all combinations of software rendering and hardware rendering, but for some reason a small while into the multiplayer window running a ROM, it crashes with "Could not open SDL sound system" or "Got a buffer underflow!". My guess is that syncing to audio or video across both emulators is causing the emulator to go into an infinite loop somewhere? The only circumstance which I can get linking to work properly is disabling both video and audio sync, which consequently causes the emulator to run free at whichever speed it can go fastest, at the cost of no audio (for me, at least) and difficult controls.

I also found that, after initially booting with both video and audio unsynced, you can resync the video (but not the audio), however it seems to sometimes hang shortly after resyncing the video. The video renderer doesn't seem to have any effects in terms of how it hangs with audio synced, however with SDL the child emulator will run for a tiny bit, then hang, and with Qt it will not start at all with just a straight hang. Starting with Sync to Video only with OpenGL causes an immediate hang (not even one frame drawn) while Qt's renderer seems to render a white screen (one frame) before hanging. Probably not related to this issue, but Qt's rendering code only seems to draw very slowly, unless you fast forward.

And on one last note, it seems that leaving the logging window with Serial IO logged just about lags out the application, although in retrospect it's probably because mgba was running unthrottled.

So in summary:

- Default configuration (sync to audio, OpenGL, SDL) hangs multiplayer child windows (and consequently the entire app) on Linux
- Disabling sync to audio and sync to video remedies the problem, regardless of renderer or audio. However, the child window persists without sound
- Enabling sync to video after the child window is created allows it to run for a short while before hanging similarly to how it hangs on loading the ROM in the child window
- The logging window is no bueno with unthrottled linking when logging Serial IO

For reference, the tested game was Pokemon Fire Red, however the issue seems to be with all games.

**Comment by endrift** [comment]
Sounds like you're building against SDL 1.2, which doesn't support multiplayer. Try building against SDL 2.

**Comment by shinyquagsire23** [comment]
I'll give that a go I suppose, I have SDL2 installed but maybe cmake didn't pick up on it.

**Comment by shinyquagsire23** [comment]
```
[maxamillion@Teh-PowerHouse build-sdl2]$ make -j8
-- Feature summary:
-- 	CLI debugger: ON
-- 	GDB stub: ON
-- 	Video recording: ON
-- 	GIF recording: ON
-- 	Screenshot/advanced savestate support: ON
-- 	ZIP support: ON
-- 	7-Zip support: ON
-- 	Better audio resampling: ON
-- Frontend summary:
-- 	Qt: ON
-- 	SDL (2): ON
-- 	Libretro core: OFF
-- 	Profiling: OFF
-- 	Test harness: OFF
-- Library summary:
-- 	Static: OFF
-- 	Shared: ON
-- Configuring done
-- Generating done
-- Build files have been written to: /home/maxamillion/workspace/mgba/build-sdl2
```

make reports SDL2, but the issues seem to be persisting.

**Comment by endrift** [comment]
Strange. Try installing Qt Multimedia and building with that enabled (it won't show up in the feature summary), then switching the audio backend in settings.

**Comment by shinyquagsire23** [comment]
Tried Qt multimedia, no difference. It does work though, outputs sound same as SDL so it's definitely there.

Just for sanity reasons though, I tried mgba's latest nightly under Windows and it seems in one way or another that's borked too with serial IO? My test case was this: Pokemon Fire Red has a multiboot "Berry Update" by pressing select+b in the title screen. So having Fire Red as player one, and player two just booting to bios, on 0.3.1 the multiboot successfully transfers as expected. The nightly does not, it just hangs. The serial loggers report both sending their outputs but it never recieves the other's input (well, rarely. I had it work once by chance, no idea how. Maybe it's a sync issue again?)

**Comment by endrift** [comment]
Huhhh, this seems to be broken for me now, too.

**Comment by endrift** [comment]
Ugh okay so apparently SDL's alsa driver doesn't let you reopen the device, meaning I'll have to do mixing myself to fix this.

**Comment by Royal-Jelly** [comment]
Well sounds like i'll have to wait for multiplayer, hope its gets resolved soon so I may once again beat the tar out of my brothers net navi

happy bug squashing sir,

**Comment by Royal-Jelly** [comment]
Im kinda disappointed that you've allowed this to become a long standing bug, you might as well remove the option for multiplayer

**Comment by endrift** [comment]
Gonna be making a lot of changes to multiplayer in 0.5, so I'll flatten it out then.

**Comment by endrift** [comment]
This might be ~fixed by 2a80ddf8d0e1eaee3437c46547e6255f039987d0, which disables audio sync for windows that can't open the audio device. That means extra players won't have sound, but will still work. Which may be desirable, as the sounds tend to stop all over each other anyway.

**Comment by endrift** [comment]
Should be properly fixed now.
